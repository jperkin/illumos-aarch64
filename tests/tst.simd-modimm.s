/*
 * This file and its contents are supplied under the terms of the
 * Common Development and Distribution License ("CDDL"), version 1.0.
 * You may only use this file in accordance with the terms of version
 * 1.0 of the CDDL.
 *
 * A full copy of the text of the CDDL should have accompanied this
 * source.  A copy of the CDDL is also available via the Internet at
 * http://www.illumos.org/license/CDDL.
 */

/*
 * Copyright 2017 Joyent, Inc.
 */

/* 8-bit */
.macro simd_modimm_8 op
	\op	v0.8b,   #0xa
	\op	v31.16b, #0x5, lsl 0
.endm
/* 16-bit shifted immediate */
.macro simd_modimm_16 op
	\op	v0.4h,  #0x5, lsl 0
	\op	v31.8h, #0xa, lsl 8
.endm
/* 32-bit shifted immediate */
.macro simd_modimm_32 op
	\op	v0.2s,  #0
	\op	v31.4s, #0x5,  lsl 8
	\op	v0.4s,  #0xa,  lsl 16
	\op	v31.2s, #0xff, lsl 24
.endm
/* 32-bit shifting ones */
.macro simd_modimm_32_1 op
	\op	v0.2s,  #0x5, msl 8
	\op	v31.4s, #0xa, msl 16
.endm
/*
 * C3.6.6 AdvSIMD modified immediate
 */
.text
.globl libdis_test
.type libdis_test, @function
libdis_test:
	simd_modimm_32 movi
	simd_modimm_32 orr
	simd_modimm_16 movi
	simd_modimm_16 orr
	simd_modimm_32_1 movi
	simd_modimm_8 movi
	/* XXX: enable when FP output fixed.
	fmov v0.2s, #2.0
	fmov v31.4s, #1.0
	*/
	simd_modimm_32 mvni
	simd_modimm_32 bic
	simd_modimm_16 mvni
	simd_modimm_16 bic
	simd_modimm_32_1 mvni
	/* scalar/vector */
	movi	d0, #0x0
	movi	d31, #0xff00ff00ff00ff00
	movi	v0.2d, #0x00ff00ff00ff00ff
	movi	v31.2d, #0xffffffffffffffff
	/* XXX: enable when FP output fixed.
	fmov v0.2d, #1.0
	fmov v31.2d, #2.0
	*/
.size libdis_test, [.-libdis_test]
