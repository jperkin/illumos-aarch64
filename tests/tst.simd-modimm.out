    libdis_test:      00 04 00 0f  movi v0.2s, #0x0
    libdis_test+0x4:  bf 24 00 4f  movi v31.4s, #0x5, lsl #8
    libdis_test+0x8:  40 45 00 4f  movi v0.4s, #0xa, lsl #16
    libdis_test+0xc:  ff 67 07 0f  movi v31.2s, #0xff, lsl #24
    libdis_test+0x10: 00 14 00 0f  orr v0.2s, #0x0
    libdis_test+0x14: bf 34 00 4f  orr v31.4s, #0x5, lsl #8
    libdis_test+0x18: 40 55 00 4f  orr v0.4s, #0xa, lsl #16
    libdis_test+0x1c: ff 77 07 0f  orr v31.2s, #0xff, lsl #24
    libdis_test+0x20: a0 84 00 0f  movi v0.4h, #0x5
    libdis_test+0x24: 5f a5 00 4f  movi v31.8h, #0xa, lsl #8
    libdis_test+0x28: a0 94 00 0f  orr v0.4h, #0x5
    libdis_test+0x2c: 5f b5 00 4f  orr v31.8h, #0xa, lsl #8
    libdis_test+0x30: a0 c4 00 0f  movi v0.2s, #0x5, msl #8
    libdis_test+0x34: 5f d5 00 4f  movi v31.4s, #0xa, msl #16
    libdis_test+0x38: 40 e5 00 0f  movi v0.8b, #0xa
    libdis_test+0x3c: bf e4 00 4f  movi v31.16b, #0x5
    libdis_test+0x40: 00 04 00 2f  mvni v0.2s, #0x0
    libdis_test+0x44: bf 24 00 6f  mvni v31.4s, #0x5, lsl #8
    libdis_test+0x48: 40 45 00 6f  mvni v0.4s, #0xa, lsl #16
    libdis_test+0x4c: ff 67 07 2f  mvni v31.2s, #0xff, lsl #24
    libdis_test+0x50: 00 14 00 2f  bic v0.2s, #0x0
    libdis_test+0x54: bf 34 00 6f  bic v31.4s, #0x5, lsl #8
    libdis_test+0x58: 40 55 00 6f  bic v0.4s, #0xa, lsl #16
    libdis_test+0x5c: ff 77 07 2f  bic v31.2s, #0xff, lsl #24
    libdis_test+0x60: a0 84 00 2f  mvni v0.4h, #0x5
    libdis_test+0x64: 5f a5 00 6f  mvni v31.8h, #0xa, lsl #8
    libdis_test+0x68: a0 94 00 2f  bic v0.4h, #0x5
    libdis_test+0x6c: 5f b5 00 6f  bic v31.8h, #0xa, lsl #8
    libdis_test+0x70: a0 c4 00 2f  mvni v0.2s, #0x5, msl #8
    libdis_test+0x74: 5f d5 00 6f  mvni v31.4s, #0xa, msl #16
    libdis_test+0x78: 00 e4 00 2f  movi d0, #0x0
    libdis_test+0x7c: 5f e5 05 2f  movi d31, #0xff00ff00ff00ff00
    libdis_test+0x80: a0 e6 02 6f  movi v0.2d, #0xff00ff00ff00ff
    libdis_test+0x84: ff e7 07 6f  movi v31.2d, #0xffffffffffffffff
